pylon-core/src/main/kotlin/io/github/pylonmc/pylon/core/item/research/Research.kt

```diff
@@ -0,0 +1,201 @@
+            return research in this.researches
+        }
+
+        @JvmStatic
+        @get:JvmName("getResearchFor")
+        val PylonItemSchema.research: Research?
+            get() = PylonRegistry.RESEARCHES.find { this.key in it.unlocks }
+
+        @JvmStatic
+        @JvmOverloads
+        @JvmName("canPlayerUse")
+        fun Player.canUse(item: PylonItemSchema, sendMessage: Boolean = false): Boolean {
+            if (
+                !PylonConfig.researchesEnabled
+                || this.gameMode == GameMode.CREATIVE
+                || this.hasPermission(item.permission)
+            ) return true
+
+            val research = item.research
+            if (research == null) return true
+
+            val canUse = this.hasResearch(research.key)
+            if (!canUse && sendMessage) {
+                val clickableResearch = research.name
+                    .hoverEvent(
+                        HoverEvent.showText(
+                            Component.translatable(
+                                "pylon.pyloncore.message.research.click_to_research",
+                                PylonArgument.of("points", Component.text(research.cost))
+                            )
+                        )
+                    )
+                    .clickEvent(ClickEvent.runCommand("/pylon research research ${research.key}"))
+                this.sendMessage(
+                    Component.translatable(
+                        "pylon.pyloncore.message.research.unknown",
+                        PylonArgument.of("item", item.itemStack.effectiveName()),
+                        PylonArgument.of("research", clickableResearch)
+                    )
+                )
+            }
+
+            return canUse
+        }
+
+        @JvmStatic
+        fun Player.clearResearches() {
+            this.researches = emptySet()
+        }
+
+        private val playerCheckerJobs = mutableMapOf<UUID, Job>()
+
+        @EventHandler(priority = EventPriority.MONITOR)
+        private fun onPlayerJoin(event: PlayerJoinEvent) {
+            if (PylonConfig.researchesEnabled) {
+                val player = event.player
+                // This task runs just in case a player manages to obtain an
+                // unknown item without picking it up somehow
+                playerCheckerJobs[player.uniqueId] = pluginInstance.launch {
+                    while (true) {
+                        player.ejectUnknownItems()
+                        delay(10.ticks)
+                    }
+                }
+            }
+        }
+
+        @EventHandler
+        private fun onPlayerPickup(event: EntityPickupItemEvent) {
+            val entity = event.entity
+            if (entity is Player) {
+                pluginInstance.launch {
+                    delay(1.ticks)
+                    entity.ejectUnknownItems()
+                }
+            }
+        }
+
+        @EventHandler
```

Comment:
```suggestion
        @EventHandler(priority = EventPriority.MONITOR)
```
should have same prio as join